### DISCUSSION TITLE
Drawing over an Embedded Process/Window

### DISCUSSION THREAD

**Discussion by BertrandDeSaintBezier** [microsoft/WindowsAppSDK#3879] [Q&A] [answered]
Hi ! 

I'm using the `Win32.dll` to embed another process/window inside my WinUI application (built using Uno Platform). This is working fine, but the issue is that sometimes, some of the UI elements (like tooltips or when you "open" a NavigationView from its compact form) will get drawn behind this embedded window. 

Is there a way to specify a Z-index for them that takes precedence over the Z-indices of the embedded window ?

Example (going from topmost to bottommost) : 

> NavigationView (drawn on left side)
>> Info panels (right and left sides)
>>> Embedded Process (center)
>>>> Application background (Rectangle with acrylic fill)

Thanks !

**Comment by DarranRowe** [comment, marked as answer]
To put it bluntly, no.
There are a couple of things at play here which works against you. First, grabbing a window and shoving it into another tends to require cooperation from both windows for everything to work correctly. There is no Xaml control or method that officially supports hosting a HWND inside a WinUI 3 application. This means that WinUI 3 is unaware of the embedded window.
Second, WinUI 3 uses an implementation of the [Visual Layer](https://learn.microsoft.com/en-us/windows/apps/desktop/modernize/visual-layer-in-desktop-apps) to do the drawing. For desktop applications, this is built upon [DirectComposition](https://learn.microsoft.com/en-us/windows/win32/directcomp/directcomposition-portal). This means that [this model](https://learn.microsoft.com/en-us/windows/win32/directcomp/basic-concepts#composition-target-window) is in use. There are two layers. The lowest layer, child windows are drawn on top of that, then finally there is an upper layer. Each layer is flattened down into a single bitmap. WinUI 3 doesn't use the upper layer though. The only way it could deal with this properly is if controls used something like Popup with IsConstrainedToRootBounds set to false which detaches it from the Visual Layer. I believe that this should create a new child HWND, but this doesn't change the fact that WinUI 3 is unaware of any other child Windows.
For this to work properly, WinUI 3 needs to support this.

Finally, WinUI 3 questions really should be in the [WinUI](https://github.com/microsoft/microsoft-ui-xaml) repository.

**Comment by castorix** [comment]
> There is no Xaml control or method that officially supports hosting a HWND inside a WinUI 3 application. 

There is _Microsoft.UI.Content.ContentExternalOutputLink_ with _IDCompositionDesktopDevice.CreateSurfaceFromHwnd_, _ElementCompositionPreview.SetElementChildVisual_ that I had tested in C# some months ago inside a XAML Image control, so drawn behind XAML controls like Popups, but for static windows (no user interaction)

**Reply by DarranRowe** [reply]
Could you provide a link to the documentation for that? I must have missed it in the 1.4 release and I can't see the documentation.

**Reply by 501586528** [reply]
I've got the same problem. And I also try to use Microsoft.UI.Content.ContentExternalOutputLink. But I don't know how to use them(Microsoft.UI.Content.ContentExternalOutputLink , IDCompositionDesktopDevice.CreateSurfaceFromHwnd, ElementCompositionPreview). Could you share your demo with me ? extremely grateful !

**Comment by castorix** [comment]
> Could you provide a link to the documentation for that? I must have missed it in the 1.4 release and I can't see the documentation.

They don't seem in docs (still marked "Experimental", I had done tests from tools like ILSpy to list all interfaces...)
